src/server/electron.ts

```diff
@@ -0,0 +1,207 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as path from 'path';
+import { CRBrowser, CRBrowserContext } from '../chromium/crBrowser';
+import { CRConnection, CRSession } from '../chromium/crConnection';
+import { CRExecutionContext } from '../chromium/crExecutionContext';
+import { TimeoutError } from '../errors';
+import { Events } from '../events';
+import { ExtendedEventEmitter } from '../extendedEventEmitter';
+import { helper } from '../helper';
+import * as js from '../javascript';
+import { InnerLogger, Logger, RootLogger } from '../logger';
+import { Page } from '../page';
+import { TimeoutSettings } from '../timeoutSettings';
+import { WebSocketTransport } from '../transport';
+import * as types from '../types';
+import { BrowserServer } from './browserServer';
+import { launchProcess, waitForLine } from './processLauncher';
+import { BrowserContext } from '../browserContext';
+
+type ElectronLaunchOptions = {
+  args?: string[],
+  cwd?: string,
+  env?: {[key: string]: string|number|boolean},
+  handleSIGINT?: boolean,
+  handleSIGTERM?: boolean,
+  handleSIGHUP?: boolean,
+  timeout?: number,
+  logger?: Logger,
+};
+
+export const ElectronEvents = {
+  ElectronApplication: {
+    Close: 'close',
+    Window: 'window',
+  }
+};
+
+export class ElectronApplication extends ExtendedEventEmitter {
+  private _logger: InnerLogger;
+  private _browserContext: CRBrowserContext;
+  private _nodeConnection: CRConnection;
+  private _nodeSession: CRSession;
+  private _nodeExecutionContext: js.ExecutionContext | undefined;
+  private _nodeElectronHandle: js.JSHandle<any> | undefined;
+  private _windows = new Set<Page>();
+  private _lastWindowId = 0;
+  readonly _timeoutSettings = new TimeoutSettings();
+
+  constructor(logger: InnerLogger, browser: CRBrowser, nodeConnection: CRConnection) {
+    super();
+    this._logger = logger;
+    this._browserContext = browser._defaultContext!;
+    this._browserContext.on(Events.BrowserContext.Close, () => this.emit(ElectronEvents.ElectronApplication.Close));
+    this._browserContext.on(Events.BrowserContext.Page, event => this._onPage(event));
+    this._nodeConnection = nodeConnection;
+    this._nodeSession = nodeConnection.rootSession;
+  }
+
+  private async _onPage(page: Page) {
```

Comment:
Unfortunately, onPage is not synchronous itself. Might be racy...

Reply:
We'll sort it out later

Reply:
Will sort it out later